include/sqlpp11/async_query_service.h

```diff
@@ -0,0 +1,118 @@
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+* ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+* ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#pragma once
+
+#ifndef SQLPP_ASYNC_QUERY_SERVICE_H
+#define SQLPP_ASYNC_QUERY_SERVICE_H
+
+#define SQLPP_ASIO_STANDALONE
+#ifdef SQLPP_ASIO_STANDALONE
+#define SQLPP_ASIO asio
+#include <asio.hpp>
+#else
+#include <boost/asio.hpp>
+#define SQLPP_ASIO boost::asio
+#endif
+
+#include <mutex>
+#include <vector>
+#include <iostream>
+#include <thread>
+#include <future>
+#include <chrono>
+#include <sqlpp11/exception.h>
+#include <sqlpp11/connection_pool.h>
+
+namespace sqlpp
+{
+	using namespace std::chrono_literals;
+	struct async_query_service
+	{
+	private:
+		struct async_io_service
+		{
+		private:
+			std::vector<std::thread> io_threads;
+			SQLPP_ASIO::io_service& _impl;
+			std::unique_ptr<SQLPP_ASIO::steady_timer> timer;
+
+			void timer_loop()
+			{
+				timer->expires_from_now(std::chrono::seconds(10));
+				timer->async_wait(std::bind(&async_io_service::timer_loop, this));
+			}
+
+		public:
+			async_io_service(SQLPP_ASIO::io_service& io_service, unsigned int thread_count)
+				: _impl(io_service)
+			{
+				timer = std::make_unique<SQLPP_ASIO::steady_timer>(_impl);
+				timer_loop(); // Keeps the io_service threads from returning
+
+				for (unsigned i = 0; i < thread_count; i++)
+				{
+					try
+					{
+						io_threads.push_back(std::move(std::thread([&] {io_service.run(); })));
+					}
+					catch (std::exception e)
+					{
+						throw sqlpp::exception("Async io service failed to create io threads.");
+					}
+				}
+			}
+
+			~async_io_service()
+			{
+				_impl.stop();
+				for (auto& thread : io_threads)
+				{
+					thread.join();
+				}
+			}
+		};
+		async_io_service _io_service;
+
+	public:
+		async_query_service(SQLPP_ASIO::io_service& io_service, unsigned int thread_count)
+			: _io_service(io_service, thread_count) {}
+
+		template<typename Connection, typename Connection_config, typename Connection_validator, typename Query, typename Bind>
+		void async_query(connection_pool<Connection, Connection_config, Connection_validator>& connection_pool, Query query, Bind callback)
+		{
+			_io_service._impl.post(
+				[&]()
+			{
+				auto async_connection = connection_pool.get_connection();
+				callback(async_connection(query));
```

Comment:
This should probably also pass `std::move(async_connection)` so that the callback can work with the async_connection. If they want to start a transaction, for instance, it would have to all take place on the same connection.